### ISSUE TITLE
OCSP Stapling (for improving HTTPS effective throughput and allow scalability)

### ISSUE THREAD

**Issue by creativeprogramming** [eclipse/vert.x#1689]
> Please read this article:
> https://letsencrypt.org/2016/10/24/squarespace-ocsp-impl.html
> 
> Please focus on:
> 
> > Our extension is a standalone library but we could equally well fold it into the netty-tcnative library itself. If there is interest, we can contribute it upstream as part of Netty’s next API-breaking development cycle.
> 
> hoping you'll show that there's interest! 

Cross posted on: https://github.com/netty/netty/issues/5943

Anyway it think in any case vertx should wrap this feature once it's ready in the HttpServer implementation, and I hope that even if Netty's will not show interst, vertex can do that and embed that implementation itself... 

It can reduce the ssl performance bottleneck that still require me to use nginx reverse proxy in front of some verxt as you know that even with tncnative we're not as good as nginx on https, maybe most of the bottleneck it's due to the missing ocsp stapling support, and it may be also a visibility opportunity for this project itself as letsencrypt can feature it too

My 2 cents

**Comment by vietj** [comment]
+1 to add this to vertx

**Comment by EmadAlblueshi** [comment]
+1

**Comment by creativeprogramming** [comment]
The developer of the JNI C code answered on https://github.com/netty/netty/issues/5943, but sadly still no answer to encourage a to share a PR. Maybe netty is not so sensible to HTTPS performance optimizations as VertX can be. Hope that someone more influent than me will contribute to encourage this collaboration in Netty or to try presenting VertX interest to the developer...

**Comment by vietj** [comment]
I think I gave my +1 to the feature on this thread

**Comment by creativeprogramming** [comment]
Yes @vietj thanks. Let's wait if something else moves.. I invited the developer to check this thread too... hope it will come here to chat...  I'm just trying to give my tiny help chatting about it, but don't worry I've a setTimeout to drop my insistence if I see I cannot help, let's wait it ;)

**Comment by creativeprogramming** [comment]
to watch for progress: https://github.com/netty/netty-tcnative/pull/198

**Comment by vietj** [comment]
thanks @creativeprogramming

**Comment by manish-panwar** [comment]
+ 1

**Comment by manish-panwar** [comment]
+1

**Comment by leogomes** [comment]
This is already implemented in Netty: https://github.com/netty/netty/tree/4.0/example/src/main/java/io/netty/example/ocsp 

:)

**Comment by creativeprogramming** [comment]
yes now Netty is ready, good news: https://github.com/netty/netty-tcnative/pull/215

**Comment by vietj** [comment]
has it been released yet ?

**Comment by leogomes** [comment]
@vietj Don't think so. It should be part of 4.1.10, I believe. But then, for Vert.x I believe you'll have to fix this one first: https://github.com/eclipse/vert.x/issues/1872

**Comment by vietj** [comment]
yeah... also need to check with grpc project and when they are going to update the version to 4.1.10.Final

**Comment by vietj** [comment]
we are planning a 3.4.2 in a few weeks, perhaps this would rather come in a 3.4.3 after when 4.1.10.Final has been released and grpc has released an upgrade using this version

**Comment by manish-panwar** [comment]
@vietj - Are we targeting it in Vertx 3.5.0 or later?

**Comment by ramtech123** [comment]
@vietj Is it still an open issue?

**Comment by vietj** [comment]
@ramtech123 I think yes

**Comment by techpavan** [comment]
I was _experimenting_ to workaround by executing the below before invoking the `listen` method.
```
SSLHelper sslHelper = ((HttpServerImpl) httpServer.getDelegate()).getSslHelper();
OpenSslServerContext sslContext1 = (OpenSslServerContext) sslHelper.getContext((VertxInternal) vertx.getDelegate());
long ctx = sslContext1.context();
if (!OpenSsl.isOcspSupported()) {
            throw new IllegalStateException("OCSP is not supported!");
}
io.netty.internal.tcnative.SSLContext.enableOcsp(ctx, false);
```
The JNI layer crashes the application with logs like below while executing the `enableOcsp` method. Any help on resolving this issue or any alternate workarounds to enable OCSP? We may not really have a clean code until Vert.x officially adds support, but if workarounds exist for immediate need, that would help.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fd41db6a673, pid=80465, tid=0x00007fd41e2b3700
#
# JRE version: OpenJDK Runtime Environment (8.0_232-b09) (build 1.8.0_232-8u232-b09-0ubuntu1~18.04.1-b09)
# Java VM: OpenJDK 64-Bit Server VM (25.232-b09 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libssl.so.1.0.0+0x3d673]  SSL_CTX_ctrl+0x243
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /opt/stash/******/hs_err_pid80465.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
```

Crash report is added here - https://gist.github.com/techpavan/bfdd49e6c0b3b0606a2856956c7123d5

Environment details:
```
Ubuntu 18.04
netty-tcnative/2.0.28.Final/linux-x86_64
Netty version: 4.1.44
Openssl version: 1.1.1d (10 Sep 2019)
Vertx version: 3.8.4
```
